operator/v1/types_ingress.go

```diff
@@ -1453,6 +1453,30 @@ type IngressControllerTuningOptions struct {
 	// +kubebuilder:validation:Optional
 	// +optional
 	MaxConnections int32 `json:"maxConnections,omitempty"`
+
+	// reloadInterval defines the minimum interval at which the router is allowed to reload
+	// to accept new changes. Increasing this value can prevent the accumulation of
+	// HAProxy processes, depending on the scenario. Increasing this interval can
+	// also lessen load imbalance on a backend's servers when using the roundrobin
+	// balancing algorithm. Alternatively, decreasing this value may decrease latency
+	// since updates to HAProxy's configuration can take effect more quickly.
+	//
+	// The value must be a time duration value; see <https://pkg.go.dev/time#ParseDuration>.
+	// Currently, the minimum value allowed is 1s, and the maximum allowed value is
+	// 720h (= 30 days). Minimum and maximum allowed values may change in future versions
```

Comment:
Why would we ever want a reload interval larger than the graceful termination time of the router pod?  If the reload interval is longer than the graceful termination, we're better off simply having the pod restart, right?  And the pod restart time is already chosen for minimal disruption.

Reply:
The grace termination period and reload interval seem like orthogonal concepts to me.  To be clear (and maybe this needs to be clarified in the godoc), `reloadInterval` controls the reload interval for the haproxy process within the container.  The router's graceful shutdown period, which is hard-coded to 45 seconds, controls the period of time that the pod will hang around and continue handling traffic when it gets a SIGTERM (presumably because the pod is being terminated).  A user might want to increase the reload interval to very high values to ensure stability during periods of time when the user anticipates high traffic loads.  During such periods of time, the user may prefer that the haproxy process not reload, and probably also would prefer that the router pod not terminate.  However, if this does happen, the user probably still wants the router have a shutdown period sufficient to drain traffic but no longer than that.

Reply:
if new routes are not served until this interval passes, then I think anything measured in more than seconds needs to have a way for the route author to know that his traffic is not being served.

Reply:
> I thought the pod termination grace period for ingress pods was set to a number of hours.

Sorry, the pod termination grace period is set to 1 hour (3600 seconds; [source](https://github.com/openshift/cluster-ingress-operator/blob/b7dbcb92d672c7ac71c47721808ce6ee3695d81a/pkg/operator/controller/ingress/deployment.go#L240-L242)).  The router's own shutdown grace period is 45 seconds [(source](https://github.com/openshift/router/blob/601ba575b7fadc1e05e7b4e2f35a660859db9788/pkg/cmd/infra/router/template.go#L701-L705)).  If a router pod ever takes much more than 45 seconds to terminate, then something is probably broken.  

> A 30 day reload interval seems excessive to me and from a practical perspective means the pod would need to be restarted. Contrasting with say 4-12h as a maximum, which still allows for change.

It does seem excessive.  However, I could see someone wanting to turn off reloads entirely during an anticipated period of peak load.  In this scenario, the user doesn't want HAProxy reloads or pod restarts.  

> if new routes are not served until this interval passes, then I think anything measured in more than seconds needs to have a way for the route author to know that his traffic is not being served.

Ideally the route status would tell the user not only whether the controller had admitted the route but also whether the HAProxy config had been updated.  That's a long-standing gap.  Right now, we tell people that they need to try connecting to the route to determine whether it's ready.  I'd like for us to cover that gap, but that would be another feature requiring some design and consideration as to how it would impact API load.

Reply:
> if new routes are not served until this interval passes, then I think anything measured in more than seconds needs to have a way for the route author to know that his traffic is not being served.

> Ideally the route status would tell the user not only whether the controller had admitted the route but also whether the HAProxy config had been updated. That's a long-standing gap. Right now, we tell people that they need to try connecting to the route to determine whether it's ready. I'd like for us to cover that gap, but that would be another feature requiring some design and consideration as to how it would impact API load.

If I'm interpreting this correctly, the PR would change the maximum amount of time for the new route to be observed from 5s to 30 days.  This changes the gap from a minor annoyance to something that could leave a user's route broken for weeks after they expect it to work.  If I have correctly understood the impact, then I think the API gap must be closed before we can add a feature that makes that gap go from seconds to weeks.

Reply:
> the PR would change the maximum amount of time for the new route to be observed from 5s to 30 days.

Only if the cluster admin chooses to increase it to 30 days.  

What is the maximum amount of time we should allow the cluster admin to increase it to?

Reply:
Hi @deads2k,
Miciah looked into various customer cases and Bugzilla reports and found that the longest value for ReloadInterval currently in use is 90s. There was 1 Bugzilla report that mentioned a ReloadInterval of 1 hour, but that came from an engineer trying to turn off reloads while testing a change in the OpenShift router. If we think that turning off reloads for testing should not be the responsibility of the ReloadInterval API field, then I think it would be fine to make the maximum-allowed interval much shorter. Do you have any suggestions for what the maximum should be? (maybe 5 to 10 minutes?)

Reply:
(I also added a warning in the godoc about setting the interval too high.)

Reply:
Following up here, the max is now 120s.